src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Documents/VisaulStudioFileChangeTracker.cs

```diff
@@ -58,71 +71,109 @@ public override void StartListening()
         {
             _foregroundDispatcher.AssertForegroundThread();
 
-            try
+            if (_fileChangeAdviseTask != null)
             {
-                if (_fileChangeCookie == VSConstants.VSCOOKIE_NIL)
-                {
-                    var hr = _fileChangeService.AdviseFileChange(
-                        FilePath,
-                        FileChangeFlags,
-                        this,
-                        out _fileChangeCookie);
-
-                    Marshal.ThrowExceptionForHR(hr);
-                }
+                // Already listening
+                return;
             }
-            catch (Exception exception)
+
+            if (_fileChangeUnadviseTask?.IsCompleted == false)
             {
-                _errorReporter.ReportError(exception);
+                // An unadvise operation is still processing, block the foreground thread until it completes.
+                _fileChangeUnadviseTask.Join();
             }
+
+            _fileChangeAdviseTask = _joinableTaskFactory.RunAsync(async () =>
+            {
+                if (_fileChangeUnadviseTask != null)
+                {
+                    // Let the unadvisement finish
+                    await _fileChangeUnadviseTask;
+                }
+
+                try
+                {
+                    return await _fileChangeService.AdviseFileChangeAsync(FilePath, FileChangeFlags, this);
+                }
+                catch (Exception exception) when (!(exception is PathTooLongException))
+                {
+                    // We can't prevent PathTooLongException's so we ignore those.
+                    _errorReporter.ReportError(exception);
+                }
+
+                return VSConstants.VSCOOKIE_NIL;
+            });
         }
 
         public override void StopListening()
         {
             _foregroundDispatcher.AssertForegroundThread();
 
-            try
+            if (_fileChangeAdviseTask == null || _fileChangeUnadviseTask?.IsCompleted == false)
             {
-                if (_fileChangeCookie != VSConstants.VSCOOKIE_NIL)
-                {
-                    var hr = _fileChangeService.UnadviseFileChange(_fileChangeCookie);
-                    Marshal.ThrowExceptionForHR(hr);
-                    _fileChangeCookie = VSConstants.VSCOOKIE_NIL;
-                }
+                // Already not listening or trying to stop listening
+                return;
             }
-            catch (Exception exception)
+
+            _fileChangeUnadviseTask = _joinableTaskFactory.RunAsync(async () =>
             {
-                _errorReporter.ReportError(exception);
-            }
+                try
+                {
+                    var fileChangeCookie = await _fileChangeAdviseTask;
```

Comment:
While `_fileChangeAdviseTask` gets awaited as part of Stop, if another `Start` is invoked, you skip the `Start` based on the check in line 74. At the end of `Stop`, you aren't listening, and the later `Start` would have done nothing. Is this something to be concerned about?